akka-http/src/main/scala/akka/http/javadsl/server/JavaScalaTypeEquivalence.scala

```diff
@@ -0,0 +1,63 @@
+package akka.http.javadsl.server
+
+import akka.http.javadsl
+import akka.http.scaladsl
+import scala.language.implicitConversions
+import scala.reflect.ClassTag
+import java.util.Optional
+
+object JavaScalaTypeEquivalence {
+  // We define implicit conversions from the javadsl model objects to the scaladsl subclasses,
+  // for all types where there is only 1 valid direct subclass of the javadsl type
+  // (being the scaladsl one)
+
+  // These are gathered here rather than doing explicit asInstanceOf throughout the code, in order
+  // to get predictable compile errors if this hierarchy might ever change.
```

Comment:
@ktoso I recall finding a similar class in the existing java/scala DSL that adds implicit `toScala` / `toJava` methods instead. I opted for direct implicit conversions here, since very often in the DSL it's not just a`T` that needs to be converted, but rather a `Seq[T]`, `Future[T]`, `Flow[T]` etc. Still, it's probably something that should be aligned.